bootstrap/app/addon-manager.go

```diff
@@ -0,0 +1,306 @@
+	}
+	return pwf
+}
+
+func (m *addonManager) Run() error {
+	m.waitForSystemServiceAccount()
+	if err := m.ensureDefaultAdmissionControlsObjects(); err != nil {
+		if os.IsNotExist(err) {
+			klog.Warningf("Admissions control object directory not found: %v", err)
+		} else {
+			klog.Errorf("Could not create admissions control objects: %v", err)
+		}
+	}
+	ticker := time.NewTicker(m.checkInterval)
+	for ; true; <-ticker.C { // Run immediately, then after every tick.
+		if !m.leaderElection || m.isLeader() {
+			m.ensureAddons()
+			m.reconcileAddons()
+		} else {
+			klog.Info("== Another addon-manager is acting as leader. Sleeping... ==")
+		}
+	}
+	return nil
+}
+
+func (m *addonManager) waitForSystemServiceAccount() {
+	var out, errOut bytes.Buffer
+	for {
+		err := m.kubectl(&out, &errOut, m.kubectlOpts, "get", "-n", systemNamespace, "serviceaccount", "default", "-o", "go-template={{with index .secrets 0}}{{.name}}{{end}}")
+		if errStr := errOut.String(); errStr != "" {
+			klog.Warning(errStr)
+		}
+		if err == nil && out.String() != "" {
+			return
+		}
+		out.Reset()
+		errOut.Reset()
+		time.Sleep(500 * time.Millisecond)
+	}
+}
+
+// Create admission_control objects if defined before any other addon services. If the limits
+// are defined in a namespace other than default, we should still create the limits for the
+// default namespace.
+func (m *addonManager) ensureDefaultAdmissionControlsObjects() error {
+	var out, errOut bytes.Buffer
+	return filepath.Walk(m.admissionControlsDir,
+		func(path string, _ os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			ext := filepath.Ext(path)
+			if ext == ".yaml" || ext == ".json" {
+				for attemptsRemaining := startAddonAttempts; attemptsRemaining > 0; attemptsRemaining-- {
+					err = m.kubectl(&out, &errOut, m.kubectlOpts, "--namespace=default", "apply", "-f", path)
+					filterLog(&out, "", klog.Info)
+					filterLog(&errOut, "", klog.Warning)
+					if err == nil {
+						klog.Infof("== Successfully started %s in namespace default ==", path)
+						return nil
+					}
+					klog.Warningf("== Failed to start %s in namespace default. %d tries remaining. ==", path, attemptsRemaining-1)
+					if attemptsRemaining > 1 {
+						time.Sleep(startAddonRetryDelay)
+					}
+				}
+			}
+			return err
+		})
+}
+
+// IsLeader returns whether this instance of addonManager should act as leader in a multi-master cluster.
+func (m *addonManager) isLeader() bool {
+	var out, errOut bytes.Buffer
+	klog.Info("== Determining addon-manager leader ==")
+	// TODO: Do a real leader election instead of piggybacking off of KCM
+	err := m.kubectl(&out, &errOut, m.kubectlOpts, "-n", systemNamespace, "get", "ep", "kube-controller-manager", "-o", `go-template={{index .metadata.annotations "control-plane.alpha.kubernetes.io/leader"}}`)
+	filterLog(&errOut, "", klog.Warning)
+	if err != nil {
+		// If the leader can't be positively established, assume that we're it
+		return true
```

Comment:
Should we log here, or is the idea that we expect something will have gone to errOut?

Reply:
> is the idea that we expect something will have gone to errOut?

right, unless we actually care about the exit code